### ISSUE TITLE
passing string arguments generates broken code

### ISSUE THREAD

**Issue by mwhudson** [4ad/go.arm64#80]
Compiling this noddy code (with -l):

``` go
package e

var s = "xxx"

func e() {
    f(s)
}

func f(s string) {

}
```

generates rather broken code:

```
"".e t=1 size=80 value=0 args=0x0 locals=0x10
    0x0000 00000 (stringarg.go:5)   TEXT    "".e+0(SB),$24-0
    0x0000 00000 (stringarg.go:5)   MOV R30,-32(R31)!
    0x0004 00004 (stringarg.go:5)   FUNCDATA    $0,gclocals·3280bececceccd33cb74587feedb1f9f+0(SB)
    0x0004 00004 (stringarg.go:5)   FUNCDATA    $1,gclocals·3280bececceccd33cb74587feedb1f9f+0(SB)
    0x0004 00004 (stringarg.go:6)   MOV $"".s+0(SB),R7
    0x0008 00008 (stringarg.go:6)   MOV R7,R7
    0x000c 00012 (stringarg.go:6)   MOV R7,R8
    0x0010 00016 (stringarg.go:6)   MOV $8,R7
    0x0014 00020 (stringarg.go:6)   MOV R7,R7
    0x0018 00024 (stringarg.go:6)   SUB $8,R8
    0x001c 00028 (stringarg.go:6)   SUB $8,R7
    0x0020 00032 (stringarg.go:6)   MOV 8(R8),R9
    0x0024 00036 (stringarg.go:6)   MOV R9,8(R7)
    0x0028 00040 (stringarg.go:6)   MOV 8(R8),R9
    0x002c 00044 (stringarg.go:6)   MOV R9,8(R7)
    0x0030 00048 (stringarg.go:6)   PCDATA  $0,$0
    0x0030 00048 (stringarg.go:6)   BL  ,"".f+0(PC)
    0x0034 00052 (stringarg.go:7)   MOV (R31)32!,R30
    0x0038 00056 (stringarg.go:7)   RET ,0(R30)
    0x003c 00060 (stringarg.go:7)   B   ,0(PC)
    0x0040 00064 (stringarg.go:7)   DWORD   ,$"".s+0(SB)
    0x0000 fe 0f 1e f8 e7 01 00 58 e7 03 07 aa e8 03 07 aa  .......X........
    0x0010 07 01 80 d2 e7 03 07 aa 08 21 00 d1 e7 20 00 d1  .........!... ..
    0x0020 09 05 40 f9 e9 04 00 f9 09 05 40 f9 e9 04 00 f9  ..@.......@.....
    0x0030 00 00 00 94 fe 07 42 f8 c0 03 5f d6 00 00 00 14  ......B..._.....
    0x0040 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
    rel 48+4 t=6 "".f+2483027968
    rel 56+0 t=7 +0
    rel 64+8 t=1 "".s+0
```

In particular, the code at offset 0x10 puts a small integer into R7 and (unless I'm blind) R7 is still a small integer (in fact, 0) when the code at offset 0x24 treats it as a pointer.

**Comment by mwhudson** [comment]
Soo... this turns out to be something like this:

agen is called for the string argument, which is a OINDREG node with n->val.u.reg of SP/R31
agen wraps its argument in a OADDR node and calls gins(AMOV, OADDR-node, ...)
gins calls naddr() on the OADDR node
this makes an D_CONST Addr with reg set REGSP
... but REGSP is also REGZERO, so this means that this D_CONST is just interpreted as a constant, not as a memory reference.

**Comment by mwhudson** [comment]
This reminds me of issue #14 in some ways -- a Addr with type D_CONST and reg == 31 is ambiguous, does it mean a constant 0 or an indirection through SP?  I guess if offset is not 0 it isn't ambiguous in this way, but ...

**Comment by 4ad** [comment]
I don't understand, where does the D_CONST come from? naddr does not set D_CONST.

**Comment by 4ad** [comment]
Can you please describe in more detail and more clearly what you think is going on? I don't understand exactly what you mean.

Regarding the const ambiguity, I don't understand what you mean either, there is no ambiguity. For constants, a->reg is usually NREG (e.g. no register). It can also be 31, but then it's always ZR. It can never be SP, and it can never be any other register. Something like 16(SP) is not D_CONST, it is D_OREG. 16(ZR) doesn't exist, so no ambiguity there. 0(R31) is Z_OREG, but its meaning is clear in the optab table. Every optab line has a known meaning of R31. When there is ambiguity in the implementation in asmout (e.g. if many optab lines map to the same case in optab), that ambiguity is solved by using (or not) REGSP in the optab param field. So there's no ambiguity there.

**Comment by 4ad** [comment]
Note that $16(SP) (e.g. effective address of dword on stack) is also D_CONST with a->reg=NREG (32), _NOT_ 31. a->name will be D_AUTO, and there is again no ambiguity there. Something like $8(R5) doesn't exist (our tools seem to accept that form too, and think it's also D_AUTO, but it's a bug, they should reject it). Again I don't see any ambiguity.

**Comment by mwhudson** [comment]
"I don't understand, where does the D_CONST come from? naddr does not set D_CONST." -- sure it does, https://github.com/4ad/go/blob/dev.arm64/src/cmd/7g/gsubr.c#L1258

" It can also be 31, but then it's always ZR. It can never be SP, and it can never be any other register." -- but REGZERO and REGSP are both 31?

Anyway to explain what is going on, I suggest: running 7g -l -g -M to compile the file I pasted above, set a breakpoint on agen and step through the second time agen is called.

**Comment by 4ad** [comment]
On Mon, Feb 2, 2015 at 7:23 PM, Michael Hudson-Doyle <
notifications@github.com> wrote:

> but REGZERO and REGSP are both 31?

Yes, so? D_CONST will never have any register set, except for ZR.

## 

Aram Hăvărneanu

**Comment by mwhudson** [comment]
Well maybe I'm chasing the wrong fish entirely, I'm happy to not debug it myself :)  You do see that the current output is completely wrong, though?

**Comment by 4ad** [comment]
Yes, it is very wrong.

**Comment by 4ad** [comment]
> D_CONST will never have any register set, except for ZR.

s/will/should

It's currently possible to generate an Addr with an arbitrary register for D_CONST, but that's clearly a bug.

**Comment by mwhudson** [comment]
This also affects passing arguments as interface{} too.  It's basically blocking further progress at this point.

**Comment by davecheney** [comment]
Hello,

my take so far this morning is passing arguments of any fat type is faulty. The compiler tries to copy the interface onto the stack for the call but ends up buggering it up completely causing a SIGBUS.

This is why it affects strings, interfaces and other fat types.

The code generated by 7g is similar to 9g, but dissimilar to 5g as 5g has a working `componentgen`.

**Comment by forsyth** [comment]
On 2 February 2015 at 21:41, Dave Cheney notifications@github.com wrote:

> my take so far this morning is passing arguments of any fat type is
> faulty. The compiler tries to copy the interface onto the stack for the
> call but ends up buggering it up completely causing a SIGBUS.

yes, it's trying to generate a block assignment and not doing too well.

**Comment by 4ad** [comment]
The problem is that a->reg should be NREG, not REGSP, and a->name should be D_AUTO (if it already isn't).

**Comment by 4ad** [comment]
One line fix, 238 LOC comment.

**Comment by 4ad** [comment]
This was a PITA to fix because while I had the fix ready in minutes, the printing code for constants still printed wrong code. It took hours until I used objdump on the binary to see that it is in fact fixed.

**Comment by 4ad** [comment]
Ok, we now have a different problem:

```
0x0020 00032 (b.go:12)  MOV 8(R1),R2
0x0024 00036 (b.go:12)  MOV R2,8(R0)
0x0028 00040 (b.go:12)  MOV 8(R1),R2
0x002c 00044 (b.go:12)  MOV R2,8(R0)
```

That should be (R1)8! and similar. sgen is broken.

**Comment by 4ad** [comment]
Fixed in 582e07f.
